src/go/overlord/overlord.go

```diff
@@ -90,11 +93,15 @@ func New(ctx context.Context, minionAddresses []string, caCertPath string) (*Ser
 	return server, nil
 }
 
+const minionInitDeadlineMs = 1000
+
 func getInterestsFromMinions(ctx context.Context, minions map[string]mpb.MinionClient) ([]*state.MappedInterest, error) {
 	var interests []*state.MappedInterest
 	for name, m := range minions {
-		// TODO(paradoxengine): most likely, a deadline here?
-		intResp, err := m.ListInitialInterests(ctx, &mpb.ListInitialInterestsRequest{})
+		// Note how we build a dedicated context for each request
+		deadlineCtx, cancel := context.WithTimeout(context.Background(), minionInitDeadlineMs*time.Millisecond)
+		defer cancel()
+		intResp, err := m.ListInitialInterests(deadlineCtx, &mpb.ListInitialInterestsRequest{})
```

Comment:
You are setting a deadline here, but you don't check it in the minions. Minions should have something like
```go
if ctx.Err() == context.Canceled {
	return status.New(codes.Canceled, "Overlord cancelled, obeying.")
}
```
Because they cover a **server** role in this call.

Moreover, this does not check how long does it take for the overall execution, but allows minions to decide whether to  respond or not, depending on how much time passed since the request was performed. I'm not sure this is what you are looking for, but if it is, please add some handling code on the other side (even with another PR is fine).

If this is not the case, please use a select statement to check for timeout.

Reply:
Actually, not :-)
I've tested at length now and here's how things work in gRPC.
The m.List call is blocking. It returns when it's done, or when the deadline is expired, in which case you get an err with the appropriate deadline expired message.

So I think the code works as it is, unless you think I should do something more?

I will add anyway the canceled check in the minions, though I don't think it's super valuable (same as the overlord)

Reply:
Actually, after looking at it I don't really feel it's worth adding the stuff on the Minions for now. It adds the dependencies, and makes the code less readable for IMO not enough advantage (given there is client-side timeout)